src/System.Slices/System/Buffers/ReferenceCounter.cs

```diff
@@ -0,0 +1,203 @@
+﻿using System.Runtime.CompilerServices;
+
+namespace System.Runtime
+{
+    /// <summary>
+    /// Make sure the struct is not copied, i.e. pass it only by reference
+    /// </summary>
+    /// <remarks>
+    /// The  counter is not completly race-free. Reading GetGlobalCount and AddReference/Release are subject to a race.
+    /// </remarks>
+    public struct ReferenceCounter
+    {
+        // thread local counts that can be updated very efficiently
+        [ThreadStatic]
+        static ObjectTable t_threadLocalCounts;
+
+        // all thread local counts; these are tallied up when global count is comupted
+        static ObjectTable[] s_allTables = new ObjectTable[4];
+        static int s_allTablesCount;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AddReference(object obj)
+        {
+            var localCounts = t_threadLocalCounts;
+            if (localCounts != null)
+            {
+                localCounts.Increment(obj);
+            }
+            else
+            {
+                AddNewObject(obj);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Release(object obj)
+        {
+            var localCounts = t_threadLocalCounts;
+            localCounts.Decrement(obj);
+        }
+
+        // TODO: can we detect if the object was only refcounted on one thread and its the current thread? If yes, we don't need to synchronize?
+        public uint GetGlobalCount(object obj)
+        {
+            var allTables = s_allTables;
+            lock (allTables)
+            {
+                uint globalCount = 0;
+                for (int index = 0; index < s_allTablesCount; index++)
+                {
+                    globalCount += allTables[index].GetCount(obj);
+                }
+                return globalCount;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public uint GetThreadLocalCount(object obj)
+        {
+            var localCounts = t_threadLocalCounts;
+            if (localCounts == null) return 0;
+            return localCounts.GetCount(obj);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static void AddNewObject(object obj)
+        {
+            var localCounts = t_threadLocalCounts;
+            lock (s_allTables)
```

Comment:
It has to be thread safe. I might try to make it use lighter weight synchronization, but the code (the array based datastructure) is already super complex. Also, this does not even happen the first time the ref count is upped. It happens the first time an object is upped. If the object is returned to the pool and then reused (while its slot is not reused), the next time the object is reused an upped from 0 to 1, it will already be in the array (which BTW might cause leaks, and so we should discuss).